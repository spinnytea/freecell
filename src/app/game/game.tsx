import { Card, RankList, shorthand, SuitList } from '@/app/game/card';

const NUMBER_OF_FREE_CELLS = 4;
const NUMBER_OF_FOUNDATIONS = SuitList.length;
const NUMBER_OF_CASCADES = 8;

export class FreeCell {
	cards: Card[];

	// structure to make the logic easier
	deck: Card[];
	cells: (Card | null)[];
	foundations: (Card | null)[];
	tableau: Card[][];

	constructor(prev?: FreeCell) {
		if (prev) {
			// cards need to remain in consitent order for react[key=""] to work
			this.cards = prev.cards.map((card) => ({ ...card }));

			this.deck = new Array<Card>(prev.deck.length);
			this.cells = new Array<null>(NUMBER_OF_FREE_CELLS).fill(null);
			this.foundations = new Array<null>(NUMBER_OF_FOUNDATIONS).fill(null);
			this.tableau = prev.tableau.map((cascade) => new Array<Card>(cascade.length));

			// we want the objects in "cards" and there rest of the game board
			this.cards.forEach((card) => {
				switch (card.location.fixture) {
					case 'deck':
						this.deck[card.location.data[0]] = card;
						break;
					case 'freecell':
						this.cells[card.location.data[0]] = card;
						break;
					case 'foundation':
						this.foundations[card.location.data[0]] = card;
						break;
					case 'cascade':
						this.tableau[card.location.data[0]][card.location.data[1]] = card;
						break;
				}
			});
		} else {
			this.cards = [];
			this.deck = [];
			this.cells = new Array<null>(NUMBER_OF_FREE_CELLS).fill(null);
			this.foundations = new Array<null>(NUMBER_OF_FOUNDATIONS).fill(null);
			this.tableau = [];
			while (this.tableau.length < NUMBER_OF_CASCADES) this.tableau.push([]);

			// initialize deck
			RankList.forEach((rank) => {
				SuitList.forEach((suit) => {
					const card: Card = {
						rank,
						suit,
						location: { fixture: 'deck', data: [this.deck.length] },
					};
					this.cards.push(card);
					this.deck.push(card);
				});
			});
		}
	}

	/**
		These deals are numbered from 1 to 32000.

		@see [Deal cards for FreeCell](https://rosettacode.org/wiki/Deal_cards_for_FreeCell)
	*/
	shuffle32(seed: number): FreeCell {
		const game = new FreeCell(this);

		if (game.deck.length !== RankList.length * SuitList.length)
			throw new Error('can only shuffle full decks');
		let i = game.deck.length;
		while (i > 0) {
			seed = (214013 * seed + 2531011) % Math.pow(2, 31);
			const idx = Math.floor(seed / Math.pow(2, 16)) % i;
			swap(game.deck, idx, i - 1);
			i--;
		}

		// update card locations
		game.deck.forEach((c, idx) => {
			c.location = { fixture: 'deck', data: [idx] };
		});

		return game;
	}

	/** @deprecated this is just for getting started; we want to animate each card delt */
	dealAll() {
		const game = new FreeCell(this);

		let c = -1;
		while (game.deck.length > 0) {
			c++;
			if (c >= NUMBER_OF_CASCADES) c = 0;
			const card = game.deck.pop();
			if (card) {
				card.location = { fixture: 'cascade', data: [c, game.tableau[c].length] };
				game.tableau[c].push(card);
			}
		}

		return game;
	}

	print() {
		let str = this.cells
			.concat(this.foundations)
			.map((card) => shorthand(card))
			.join(' ');
		const max = Math.max(...this.tableau.map((cascade) => cascade.length));
		for (let i = 0; i < max; i++) {
			const line = [];
			for (let c = 0; c < NUMBER_OF_CASCADES; c++) {
				line.push(shorthand(this.tableau[c][i]));
			}
			str += '\n' + line.join(' ');
		}
		if (this.deck.length) {
			str +=
				'\nd' +
				this.deck
					.slice(0)
					.reverse()
					.map((card) => shorthand(card))
					.join(' ');
		}
		return str;
	}
}

// FIXME remove
function swap<T>(array: T[], i: number, j: number) {
	const temp = array[i];
	array[i] = array[j];
	array[j] = temp;
}
