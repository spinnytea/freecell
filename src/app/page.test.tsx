import { fireEvent, render, screen } from '@testing-library/react';
import { gsap } from 'gsap/all';
import { FreeCell } from '@/app/game/game';
import GameBoard from '@/app/GameBoard';
import { StaticFixtureSizesContextProvider } from '@/app/hooks/contexts/FixtureSizes/StaticFixtureSizesContextProvider';
import StaticGameContextProvider from '@/app/hooks/contexts/Game/StaticGameContextProvider';
import { ManualTestingSettingsContextProvider } from '@/app/hooks/contexts/Settings/ManualTestingSettingsContextProvider';
import { ErrorBoundary } from '@/app/hooks/ErrorBoundary';

jest.mock('gsap/all', () => ({
	gsap: {
		to: () => ({}),
		set: () => ({}),
		timeline: () => ({}),
		utils: {
			random: () => undefined,
		},
	},
}));

function MockGamePage({ game }: { game: FreeCell }) {
	return (
		<ErrorBoundary>
			<ManualTestingSettingsContextProvider>
				<StaticGameContextProvider games={[game]}>
					<StaticFixtureSizesContextProvider>
						<GameBoard className="none" />
					</StaticFixtureSizesContextProvider>
				</StaticGameContextProvider>
			</ManualTestingSettingsContextProvider>
		</ErrorBoundary>
	);
}

describe('Free Cell UI', () => {
	// REVIEW (techdebt) this is a _lot_ of mocking, duplicated in useCardPositionAnimations.test
	let toGsapSpy: jest.SpyInstance;
	let setGsapSpy: jest.SpyInstance;
	let fromToSpy: jest.SpyInstance;
	let toSpy: jest.SpyInstance;
	let setSpy: jest.SpyInstance;
	let addLabelSpy: jest.SpyInstance;
	let addSpy: jest.SpyInstance;
	let timeScaleSpy: jest.SpyInstance;
	let timelineOnComplete: gsap.Callback | undefined;
	let consoleDebugSpy: jest.SpyInstance;
	beforeEach(() => {
		toGsapSpy = jest.spyOn(gsap, 'to');
		setGsapSpy = jest.spyOn(gsap, 'set');
		fromToSpy = jest.fn();
		toSpy = jest.fn();
		setSpy = jest.fn();
		addLabelSpy = jest.fn();
		addSpy = jest.fn();
		timeScaleSpy = jest.fn();

		jest.spyOn(gsap, 'timeline').mockImplementation((vars) => {
			timelineOnComplete = vars?.onComplete;
			const timelineMock: unknown = {
				fromTo: fromToSpy,
				to: toSpy,
				set: setSpy,
				addLabel: addLabelSpy,
				add: addSpy,
				timeScale: timeScaleSpy,
				totalProgress: () => ({
					kill: () => {
						/* empty */
					},
				}),
			};
			return timelineMock as gsap.core.Timeline;
		});
		consoleDebugSpy = jest.spyOn(console, 'debug').mockImplementation();
	});

	function mockReset(runOnComplete = true) {
		if (timelineOnComplete && runOnComplete) {
			timelineOnComplete();
		}

		toGsapSpy.mockReset();
		setGsapSpy.mockReset();
		toSpy.mockReset();
		fromToSpy.mockReset();
		setSpy.mockReset();
		addLabelSpy.mockReset();
		addSpy.mockReset();
		timeScaleSpy.mockReset();
	}

	// just pick a seed for an easy game
	//  - click to move single
	//  - click to move sequence
	//  - click to move to/from cell
	//  - click to send home
	//  - some auto-plays are fine
	test.todo('we can win a game by clicking through cards');

	/** https://www.solitairelaboratory.com/tutorial.html */
	test('renders a game', () => {
		const { container } = render(<MockGamePage game={new FreeCell().shuffle32(0)} />);

		// initial state
		expect(toGsapSpy).toHaveBeenCalledTimes(52);
		expect(setGsapSpy).toHaveBeenCalledTimes(52);
		expect(toSpy).toHaveBeenCalledTimes(0);
		expect(fromToSpy).toHaveBeenCalledTimes(0);
		expect(setSpy).toHaveBeenCalledTimes(52);
		expect(addLabelSpy.mock.calls).toEqual([['updateCardPositions']]);
		expect(addSpy).toHaveBeenCalledTimes(0);
		expect(timeScaleSpy).toHaveBeenCalledTimes(0);
		expect(consoleDebugSpy.mock.calls).toEqual([]);

		mockReset();
		fireEvent.click(screen.getAllByAltText('card back')[0]);

		// animations
		expect(toGsapSpy).toHaveBeenCalledTimes(0);
		expect(setGsapSpy).toHaveBeenCalledTimes(0);
		expect(toSpy).toHaveBeenCalledTimes(52);
		expect(fromToSpy).toHaveBeenCalledTimes(52);
		expect(setSpy).toHaveBeenCalledTimes(0);
		expect(addLabelSpy.mock.calls).toEqual([['updateCardPositions']]);
		expect(addSpy).toHaveBeenCalledTimes(0);
		expect(timeScaleSpy).toHaveBeenCalledTimes(0);
		expect(consoleDebugSpy.mock.calls).toEqual([]);

		expect(container).toMatchSnapshot();
	});

	// FIXME test.todo, reimpl moveByShorthand into clicks(screen)
	/** https://www.solitairelaboratory.com/tutorial.html */
	test.todo('Game #5 (tutorial)');
});
