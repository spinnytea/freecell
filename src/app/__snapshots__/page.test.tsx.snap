// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Free Cell UI renders a game 1`] = `
<div>
  <main
    class="none common"
    tabindex="0"
  >
    <div
      class="hiddenDeckBack"
    >
      <img
        alt="card back"
        data-nimg="1"
        decoding="async"
        draggable="false"
        fetchpriority="high"
        height="0"
        src="/i/Card_back_10.svg"
        style="color: transparent;"
        width="0"
      />
    </div>
    <div
      id="piles"
    >
      <div
        class="pile cursorPile"
        id="pilemarker-a-1-0"
        style="top: 6px; left: 11px; width: 8px; height: 18px;"
      />
      <div
        class="pile"
        id="pilemarker-b-2-0"
        style="top: 6px; left: 21px; width: 8px; height: 18px;"
      />
      <div
        class="pile"
        id="pilemarker-c-3-0"
        style="top: 6px; left: 31px; width: 8px; height: 18px;"
      />
      <div
        class="pile"
        id="pilemarker-d-4-0"
        style="top: 6px; left: 41px; width: 8px; height: 18px;"
      />
      <div
        class="pile"
        id="pilemarker-h-1-0"
        style="top: 6px; left: 51px; width: 8px; height: 18px;"
      />
      <div
        class="pile"
        id="pilemarker-h-2-0"
        style="top: 6px; left: 61px; width: 8px; height: 18px;"
      />
      <div
        class="pile"
        id="pilemarker-h-3-0"
        style="top: 6px; left: 71px; width: 8px; height: 18px;"
      />
      <div
        class="pile"
        id="pilemarker-h-4-0"
        style="top: 6px; left: 81px; width: 8px; height: 18px;"
      />
      <div
        class="pile"
        id="pilemarker-1-1-0"
        style="top: 21px; left: 11px; width: 8px; height: 18px;"
      />
      <div
        class="pile"
        id="pilemarker-2-2-0"
        style="top: 21px; left: 21px; width: 8px; height: 18px;"
      />
      <div
        class="pile"
        id="pilemarker-3-3-0"
        style="top: 21px; left: 31px; width: 8px; height: 18px;"
      />
      <div
        class="pile"
        id="pilemarker-4-4-0"
        style="top: 21px; left: 41px; width: 8px; height: 18px;"
      />
      <div
        class="pile"
        id="pilemarker-5-5-0"
        style="top: 21px; left: 51px; width: 8px; height: 18px;"
      />
      <div
        class="pile"
        id="pilemarker-6-6-0"
        style="top: 21px; left: 61px; width: 8px; height: 18px;"
      />
      <div
        class="pile"
        id="pilemarker-7-7-0"
        style="top: 21px; left: 71px; width: 8px; height: 18px;"
      />
      <div
        class="pile"
        id="pilemarker-8-8-0"
        style="top: 21px; left: 81px; width: 8px; height: 18px;"
      />
    </div>
    <div
      id="cards"
    >
      <div
        class="card"
        data-left="60"
        data-top="20"
        data-zindex="0"
        id="cAC-0"
      >
        <img
          alt="ace of clubs"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/ace_of_clubs.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="10"
        data-top="25"
        data-zindex="5"
        id="cAD-0"
      >
        <img
          alt="ace of diamonds"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/ace_of_diamonds.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="80"
        data-top="23"
        data-zindex="3"
        id="cAH-0"
      >
        <img
          alt="ace of hearts"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/ace_of_hearts.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="20"
        data-top="25"
        data-zindex="5"
        id="cAS-0"
      >
        <img
          alt="ace of spades"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/ace_of_spades.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="70"
        data-top="23"
        data-zindex="3"
        id="c2C-0"
      >
        <img
          alt="2 of clubs"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/2_of_clubs.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="30"
        data-top="22"
        data-zindex="2"
        id="c2D-0"
      >
        <img
          alt="2 of diamonds"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/2_of_diamonds.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="30"
        data-top="26"
        data-zindex="6"
        id="c2H-0"
      >
        <img
          alt="2 of hearts"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/2_of_hearts.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="10"
        data-top="26"
        data-zindex="6"
        id="c2S-0"
      >
        <img
          alt="2 of spades"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/2_of_spades.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="60"
        data-top="21"
        data-zindex="1"
        id="c3C-0"
      >
        <img
          alt="3 of clubs"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/3_of_clubs.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="80"
        data-top="22"
        data-zindex="2"
        id="c3D-0"
      >
        <img
          alt="3 of diamonds"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/3_of_diamonds.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="70"
        data-top="24"
        data-zindex="4"
        id="c3H-0"
      >
        <img
          alt="3 of hearts"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/3_of_hearts.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="80"
        data-top="21"
        data-zindex="1"
        id="c3S-0"
      >
        <img
          alt="3 of spades"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/3_of_spades.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="20"
        data-top="26"
        data-zindex="6"
        id="c4C-0"
      >
        <img
          alt="4 of clubs"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/4_of_clubs.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="30"
        data-top="25"
        data-zindex="5"
        id="c4D-0"
      >
        <img
          alt="4 of diamonds"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/4_of_diamonds.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="70"
        data-top="25"
        data-zindex="5"
        id="c4H-0"
      >
        <img
          alt="4 of hearts"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/4_of_hearts.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="30"
        data-top="21"
        data-zindex="1"
        id="c4S-0"
      >
        <img
          alt="4 of spades"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/4_of_spades.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="50"
        data-top="22"
        data-zindex="2"
        id="c5C-0"
      >
        <img
          alt="5 of clubs"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/5_of_clubs.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="70"
        data-top="21"
        data-zindex="1"
        id="c5D-0"
      >
        <img
          alt="5 of diamonds"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/5_of_diamonds.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="20"
        data-top="20"
        data-zindex="0"
        id="c5H-0"
      >
        <img
          alt="5 of hearts"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/5_of_hearts.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="40"
        data-top="26"
        data-zindex="6"
        id="c5S-0"
      >
        <img
          alt="5 of spades"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/5_of_spades.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="80"
        data-top="20"
        data-zindex="0"
        id="c6C-0"
      >
        <img
          alt="6 of clubs"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/6_of_clubs.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="30"
        data-top="23"
        data-zindex="3"
        id="c6D-0"
      >
        <img
          alt="6 of diamonds"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/6_of_diamonds.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="10"
        data-top="21"
        data-zindex="1"
        id="c6H-0"
      >
        <img
          alt="6 of hearts"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/6_of_hearts.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="50"
        data-top="20"
        data-zindex="0"
        id="c6S-0"
      >
        <img
          alt="6 of spades"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/6_of_spades.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="80"
        data-top="25"
        data-zindex="5"
        id="c7C-0"
      >
        <img
          alt="7 of clubs"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/7_of_clubs.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="70"
        data-top="22"
        data-zindex="2"
        id="c7D-0"
      >
        <img
          alt="7 of diamonds"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/7_of_diamonds.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="10"
        data-top="24"
        data-zindex="4"
        id="c7H-0"
      >
        <img
          alt="7 of hearts"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/7_of_hearts.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="50"
        data-top="23"
        data-zindex="3"
        id="c7S-0"
      >
        <img
          alt="7 of spades"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/7_of_spades.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="40"
        data-top="22"
        data-zindex="2"
        id="c8C-0"
      >
        <img
          alt="8 of clubs"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/8_of_clubs.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="40"
        data-top="24"
        data-zindex="4"
        id="c8D-0"
      >
        <img
          alt="8 of diamonds"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/8_of_diamonds.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="20"
        data-top="24"
        data-zindex="4"
        id="c8H-0"
      >
        <img
          alt="8 of hearts"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/8_of_hearts.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="40"
        data-top="25"
        data-zindex="5"
        id="c8S-0"
      >
        <img
          alt="8 of spades"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/8_of_spades.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="60"
        data-top="23"
        data-zindex="3"
        id="c9C-0"
      >
        <img
          alt="9 of clubs"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/9_of_clubs.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="10"
        data-top="23"
        data-zindex="3"
        id="c9D-0"
      >
        <img
          alt="9 of diamonds"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/9_of_diamonds.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="20"
        data-top="23"
        data-zindex="3"
        id="c9H-0"
      >
        <img
          alt="9 of hearts"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/9_of_hearts.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="50"
        data-top="25"
        data-zindex="5"
        id="c9S-0"
      >
        <img
          alt="9 of spades"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/9_of_spades.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="40"
        data-top="20"
        data-zindex="0"
        id="cTC-0"
      >
        <img
          alt="10 of clubs"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/10_of_clubs.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="10"
        data-top="22"
        data-zindex="2"
        id="cTD-0"
      >
        <img
          alt="10 of diamonds"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/10_of_diamonds.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="10"
        data-top="20"
        data-zindex="0"
        id="cTH-0"
      >
        <img
          alt="10 of hearts"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/10_of_hearts.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="70"
        data-top="20"
        data-zindex="0"
        id="cTS-0"
      >
        <img
          alt="10 of spades"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/10_of_spades.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="40"
        data-top="23"
        data-zindex="3"
        id="cJC-0"
      >
        <img
          alt="jack of clubs"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/jack_of_clubs2.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="40"
        data-top="21"
        data-zindex="1"
        id="cJD-0"
      >
        <img
          alt="jack of diamonds"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/jack_of_diamonds2.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="80"
        data-top="24"
        data-zindex="4"
        id="cJH-0"
      >
        <img
          alt="jack of hearts"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/jack_of_hearts2.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="50"
        data-top="21"
        data-zindex="1"
        id="cJS-0"
      >
        <img
          alt="jack of spades"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/jack_of_spades2.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="60"
        data-top="24"
        data-zindex="4"
        id="cQC-0"
      >
        <img
          alt="queen of clubs"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/queen_of_clubs2.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="20"
        data-top="21"
        data-zindex="1"
        id="cQD-0"
      >
        <img
          alt="queen of diamonds"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/queen_of_diamonds2.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="20"
        data-top="22"
        data-zindex="2"
        id="cQH-0"
      >
        <img
          alt="queen of hearts"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/queen_of_hearts2.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="60"
        data-top="22"
        data-zindex="2"
        id="cQS-0"
      >
        <img
          alt="queen of spades"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/queen_of_spades2.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="50"
        data-top="24"
        data-zindex="4"
        id="cKC-0"
      >
        <img
          alt="king of clubs"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/king_of_clubs2.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="60"
        data-top="25"
        data-zindex="5"
        id="cKD-0"
      >
        <img
          alt="king of diamonds"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/king_of_diamonds2.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="30"
        data-top="24"
        data-zindex="4"
        id="cKH-0"
      >
        <img
          alt="king of hearts"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/king_of_hearts2.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
      <div
        class="card"
        data-left="30"
        data-top="20"
        data-zindex="0"
        id="cKS-0"
      >
        <img
          alt="king of spades"
          data-nimg="1"
          decoding="async"
          draggable="false"
          fetchpriority="high"
          height="14"
          src="http://localhost/i/SVG-cards-1.3/king_of_spades2.svg"
          style="color: transparent;"
          width="10"
        />
      </div>
    </div>
    <button
      aria-label="Open settings dialog"
      class="btn square settingsButton"
      title="Settings"
    >
      <span
        class="squareText"
      >
        ⚙
        ︎
      </span>
    </button>
    <button
      aria-label="Undo previous action"
      class="btn square undoButton"
      title="Undo"
    >
      <span
        class="squareText"
      >
        ⎌
      </span>
    </button>
    <section
      class="status"
    >
      <label
        class="checkbox"
        for="showDebugInfo"
      >
        <input
          id="showDebugInfo"
          name="showDebugInfo"
          type="checkbox"
        />
        <span>
          Show Debug Info
        </span>
      </label>
      <span
        class="statusspacer"
      />
      <a
        href="/manualtesting"
      >
        ↗ Manual Testing
      </a>
      <span>
        vUnknown
      </span>
    </section>
  </main>
</div>
`;
